packages/framework/src/client.ts

```diff
@@ -95,21 +97,49 @@ export class Client {
     return builtConfiguration;
   }
 
+  /**
+   * Adds workflows to the client.
+   *
+   * A locking mechanism is used to ensure that duplicate workflows are not added.
+   *
+   * @param workflows - The workflows to add.
+   */
   public async addWorkflows(workflows: Array<Workflow>): Promise<void> {
     for (const workflow of workflows) {
-      if (this.discoveredWorkflows.some((existing) => existing.workflowId === workflow.id)) {
-        return;
-      } else {
-        const definition = await workflow.discover();
-        prettyPrintDiscovery(definition);
-        this.discoveredWorkflows.push(definition);
+      if (this.discoveredWorkflows.has(workflow.id)) {
+        continue;
+      }
+
+      if (this.addingWorkflows.has(workflow.id)) {
+        // Wait for the existing promise to resolve if the workflow is already being added
+        await this.workflowPromises.get(workflow.id);
+        continue;
       }
+
+      this.addingWorkflows.add(workflow.id);
+
+      const workflowPromise = this.addWorkflow(workflow);
+      this.workflowPromises.set(workflow.id, workflowPromise);
+
+      await workflowPromise;
+    }
+  }
+
+  private async addWorkflow(workflow: Workflow): Promise<void> {
+    try {
+      const definition = await workflow.discover();
```

Comment:
Good catch! With an async discover and transform this could become more prominent.